pkg/storage/tpr/storage_interface.go

```diff
@@ -215,23 +215,29 @@ func (t *store) Watch(
 		glog.Errorf("initiating the raw watch (%s)", err)
 		return nil, err
 	}
-	filteredIFace := watch.Filter(watchIface, watchFilterer(t, ns))
+	filteredIFace := watch.Filter(watchIface, watchFilterer(t, ns, false))
 	return filteredIFace, nil
 }
 
-func watchFilterer(t *store, ns string) func(watch.Event) (watch.Event, bool) {
+// watchFilterer returns a function that can be used as an argument to watch.Filter
+func watchFilterer(t *store, ns string, list bool) func(watch.Event) (watch.Event, bool) {
 	return func(in watch.Event) (watch.Event, bool) {
 		encodedBytes, err := runtime.Encode(t.codec, in.Object)
 		if err != nil {
 			glog.Errorf("couldn't encode watch event object (%s)", err)
 			return watch.Event{}, false
 		}
-		finalObj := t.singularShell("", "")
+		var finalObj runtime.Object
+		if !list {
+			finalObj = t.singularShell("", "")
+		} else {
+			finalObj = t.listShell()
+		}
 		if err := decode(t.codec, encodedBytes, finalObj); err != nil {
 			glog.Errorf("couldn't decode watch event bytes (%s)", err)
 			return watch.Event{}, false
 		}
-		if !t.hasNamespace {
+		if !t.hasNamespace && !list {
```

Comment:
What if it _is_ a list? Shouldn't we iterate over the list items and strip the namespace the way we do in `List()`?